### PR TITLE
#517 - added basic facebook login.

### PR DESCRIPTION
**Pull request by aviklai**
Still need to:
1. Improve login UI
2. Add email and password login.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -42,6 +43,12 @@
 app.config['SECURITY_REGISTERABLE'] = False
 app.config['SECURITY_USER_IDENTITY_ATTRIBUTES'] = 'username'
 app.config['BABEL_DEFAULT_LOCALE'] = 'he'
+app.config['OAUTH_CREDENTIALS'] = {
```
Comment (danielhers) [inline main.py]: These should be environment variables that we'll set on the server.
You can't save a secret code in the code when it's open source.

**Comment by aviklai** [comment]
I knew I forgot to remove someting... sorry...

File: models.py [hunk 2]
```diff
@@ -36,9 +36,11 @@ class User(Base, UserMixin):
     password = Column(String(256))
     active = Column(Boolean())
     confirmed_at = Column(DateTime())
+    social_id = Column(String(64), nullable=True, unique=True)
```
Comment (danielhers) [inline models.py]: This will require a migration. You should create one with Alembic and put it under `alembic/versions`.

File: static/js/app.js [hunk 3]
```diff
@@ -749,6 +753,11 @@ $(function () {
         getDiscussionUrl: function (identifier) {
             return "/?discussion=" + identifier + "&" + app.getCurrentUrlParams();
         },
+        LoginDialogLoad: function(){
```
Comment (danielhers) [inline static/js/app.js]: Function name should start with lower case.

File: static/js/app.js [hunk 4]
```diff
@@ -749,6 +753,11 @@ $(function () {
...
+            if (this.createDialog) this.createDialog.close();
```
Comment (danielhers) [inline static/js/app.js]: After it's closed, will this `if`'s condition be false? How?

File: static/js/logindialog.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+	className: "login-dialog",
+    events: {
+		"submit" : "submit",
+		"click .close-button" : "close"
+	},
+	render: function() {
+		this.$el.append($("#login-dialog-template").html());
+		$(document.body).append(this.$el);
+		this.modal = this.$el.find(".modal");
+        this.modal.modal("show");
```
Comment (danielhers) [inline static/js/logindialog.js]: Fix tabs in the file to use spaces instead.

File: templates/index.html [hunk 6]
```diff
@@ -40,8 +40,14 @@
 <div id="app">
     <div class="navbar navbar-fixed-top">
         <div class="navbar-inner" id="step1Component" style="vertical-align:middle; line-height: 40px; {{ '; display: none' if map_only }}">
-            <div style="float: left; visibility: hidden">
-                <span style="border-right: 1px solid #ddd; margin-left: 8px; padding-right: 8px; cursor: pointer">התחבר</span>
+            <div style="float: left;">
+                <span id="connection_status" style="border-right: 1px solid #ddd; margin-left: 8px; padding-right: 8px; cursor: pointer" onclick="app.LoginDialogLoad()">
+                    {% if current_user.is_authenticated %}
+                        Hi {{ current_user.nickname }}
```
Comment (danielhers) [inline templates/index.html]: Messages should be translated, now that we have support for that. Use `gettext` and add the translation according to this: https://github.com/hasadna/anyway/blob/dev/TRANSLATE.md

**Comment by danielhers** [comment]
Thanks, this is something that will be good to have.
You're right about the UI needing improvement. I think Facebook give a small "login with Facebook" that we can use instead of the text. Shouldn't be too hard.
Also, when I tried to sign in to the PR staging app, it went to [this URL](https://graph.facebook.com/oauth/authorize?scope=email&redirect_uri=http%3A%2F%2Fanyway-unstable-pr-541.herokuapp.com%2Fcallback%2Ffacebook&response_type=code&client_id=362934167237010) and failed with this error:

    {
       "error": {
          "message": "Invalid redirect_uri: Given URL is not allowed by the Application configuration",
          "type": "OAuthException",
          "code": 191,
          "fbtrace_id": "FHZlQSYkUnr"
       }
    }

File: static/js/app.js (hunk 4 above)
Comment (aviklai) [inline static/js/app.js]: What do you mean? if a dialog is already opened, it is supposed to close it.

**Comment by aviklai** [comment]
I got the same error when I didn't have facebook id and secret key environment variables.
Maybe they are not configured for the staging app, like they are configured for the production app?

File: models.py (hunk 2 above)
Comment (aviklai) [inline models.py]: O.K, I added a manually made migration using Alembic.

**Comment by aviklai** [comment]
@danielhers How do we proceed?

**Comment by danielhers** [comment]
I'll review this tonight or tomorrow

**Comment by danielhers** [comment]
I was trying to run the migration but Alembic thinks it's currently at the initial version, and tries to run the first migration that is already present in the DB.
For this reason or another, currently trying to log in with Facebook on http://anyway-unstable.herokuapp.com/ results in an error:

    {
       "error": {
          "message": "Invalid redirect_uri: Given URL is not allowed by the Application configuration",
          "type": "OAuthException",
          "code": 191,
          "fbtrace_id": "Ad9tzrXfe9L"
       }
    }

Need to set the current version in the [DB](https://postgres.heroku.com/databases/anyway-unstable-database) as per the instructions in http://alembic.readthedocs.org/en/latest/tutorial.html

When merging to `master` we'll have to repeat the process.

Also, anyone running Anyway locally also has to run the migration (or delete their `local.db` and re-create it).

**Comment by danielhers** [comment]
Fixed. I ran the migration and edited the website URL in the Facebook settings to be `http://anyway-unstable.herokuapp.com/`. We'll need to set this to the production site's URL when this feature moves to production.

**Comment by aviklai** [comment]
Great! By the way, I have read that the reason it works with localhost address is that facebook addressed this issue so developers could check the login from localhost (with the real app domain). That is the reason it worked on localhost.

**Comment by aviklai** [comment]
And your change won't affect the like and share functionality on the production site, right?

**Comment by danielhers** [comment]
I checked and the like/share on production still work.

**Pull request merged**